hal/src/nRF52840/usart_hal.cpp

```diff
@@ -221,54 +221,81 @@ class Usart {
         nrf_uarte_enable(uarte_);
 
         config_ = conf;
-        enabled_ = true;
+        state_ = HAL_USART_STATE_ENABLED;
 
         enableTimer();
         startReceiver();
 
-        return 0;
+        return SYSTEM_ERROR_NONE;
     }
 
     int end() {
-        CHECK_TRUE(enabled_, SYSTEM_ERROR_INVALID_STATE);
+        CHECK_TRUE(state_ != HAL_USART_STATE_DISABLED, SYSTEM_ERROR_INVALID_STATE);
         AtomicSection lk;
 
-        disableInterrupts();
-
-        stopTransmission();
-        stopReceiver();
-
-        nrf_uarte_disable(uarte_);
+        CHECK(disable());
 
-        nrf_uarte_txrx_pins_disconnect(uarte_);
-        nrf_uarte_hwfc_pins_disconnect(uarte_);
-
-        HAL_Pin_Mode(txPin_, INPUT);
+        // Configuring the pin mode to PIN_MODE_NONE will disconnect input buffer to enable power savings
+        HAL_Pin_Mode(txPin_, PIN_MODE_NONE);
         HAL_Set_Pin_Function(txPin_, PF_NONE);
-        HAL_Pin_Mode(rxPin_, INPUT);
+        HAL_Pin_Mode(rxPin_, PIN_MODE_NONE);
         HAL_Set_Pin_Function(rxPin_, PF_NONE);
-
-        if (config_.config & SERIAL_FLOW_CONTROL_CTS) {
-            HAL_Pin_Mode(ctsPin_, INPUT);
-            HAL_Set_Pin_Function(ctsPin_, PF_NONE);
-        }
         if (config_.config & SERIAL_FLOW_CONTROL_RTS) {
-            HAL_Pin_Mode(rtsPin_, INPUT);
+            HAL_Pin_Mode(rtsPin_, PIN_MODE_NONE);
             HAL_Set_Pin_Function(rtsPin_, PF_NONE);
         }
+        if (config_.config & SERIAL_FLOW_CONTROL_CTS) {
+            HAL_Pin_Mode(ctsPin_, PIN_MODE_NONE);
+            HAL_Set_Pin_Function(ctsPin_, PF_NONE);
+        }
 
-        nrfx_prs_release(uarte_);
+        state_ = HAL_USART_STATE_DISABLED;
+        transmitting_ = false;
+        receiving_ = 0;
+        config_ = {};
+        rxBuffer_.reset();
+        txBuffer_.reset();
 
-        disableTimer();
+        return SYSTEM_ERROR_NONE;
+    }
 
-        enabled_ = false;
-        config_ = {};
+    int suspend() {
+        CHECK_TRUE(isEnabled(), SYSTEM_ERROR_INVALID_STATE);
+
+        flush();
+
+        AtomicSection lk;
+
+        // Update current available received data
+        data();
+        
+        CHECK(disable());
+
+        // Configuring the input pins' mode to PIN_MODE_NONE will disconnect input buffer to enable power savings
+        // Configuring the output pins' mode to INPUT_PULLUP to not polluting the state on other side.
+        // NOTE: Do not reset pin function!
+        HAL_Pin_Mode(txPin_, INPUT_PULLUP);
+        HAL_Pin_Mode(rxPin_, PIN_MODE_NONE);
+        if (config_.config & SERIAL_FLOW_CONTROL_RTS) {
```

Comment:
Answer to https://github.com/particle-iot/device-os/pull/2125/files?file-filters%5B%5D=.c&file-filters%5B%5D=.cpp#r451252956

Reply:
This is still somewhat of a race condition. Ideally the steps should be to de-assert RTS before performing any deinitialization so that the other side again stops sending data before start doing anything.

Reply:
Probably fine as-is for now.

Reply:
Will refactor to change the sequence.